src/analysis/indices/anomaly_index.py

```diff
@@ -0,0 +1,139 @@
+    """
+    name = 'rainfall_anomaly_index'
+
+    @staticmethod
+    def assign_magnitudes(y: xr.Dataset,
+                          anom: xr.DataArray,
+                          rai_plus: xr.DataArray,
+                          rai_minus: xr.DataArray,
+                          variable: str) -> xr.Dataset:
+        """ assign magnitudes to positive and negative anomalies """
+        # REVERSE boolean indexing
+        # if anom values are NOT less than/equal than fill with PLUS values
+        y_plus = y.where(~(anom >= 0), other=rai_plus)
+        y_plus_mask = y.where(~(anom >= 0)).isnull()
+        y_plus = y_plus.where(y_plus_mask)
+
+        y_minus = y.where(~(anom < 0), other=rai_minus)
+        y_minus_mask = y.where(~(anom < 0)).isnull()
+        y_minus = y_minus.where(y_minus_mask)
+
+        # convert to numpy masked array
+        y_plus_m_np = y_plus_mask[variable].values
+        y_plus_np = y_plus[variable].values
+        y_plus_np = np.ma.array(y_plus_np, mask=(~y_plus_m_np))
+
+        y_minus_m_np = y_minus_mask[variable].values
+        y_minus_np = y_minus[variable].values
+        y_minus_np = np.ma.array(y_minus_np, mask=(~y_minus_m_np))
+
+        # recombine masked arrays
+        rai_array = (
+            np.ma.array(
+                y_plus_np.filled(1) * y_minus_np.filled(1),
+                mask=(y_plus_np.mask * y_minus_np.mask)
+            )
+        )
+
+        rai = xr.ones_like(y)
+        rai[variable] = (['time', 'lat', 'lon'], rai_array)
+        return rai
+
+    def RAI(self, da: xr.DataArray,
+            variable: str,
+            dim: str = 'time',) -> xr.DataArray:
+        """ Rainfall Anomaly Index """
+        # calculations
+        y = da.copy()
+        x1 = da.copy().sortby(dim, ascending=False)  # high -> low over TIME
+
+        # sample average, max average, min average
+        x_avg = x1.mean(dim=dim)  # monthly average over all time
+        mx_avg = x1.isel(time=slice(0, 10)).mean(dim=dim)  # max mean (top 10)
+        mn_avg = x1.isel(time=slice(-11, -1)).mean(dim=dim)  # min mean (bottom 10) NB
+
+        # anomaly = value - mean
+        anom = da - x_avg
+
+        # where anomaly is >=0, get the ratio to difference of max_avg - x_avg
+        rai_plus = 3.0 * anom.where(anom >= 0) / (mx_avg - x_avg)
+        rai_minus = -3.0 * anom.where(anom < 0) / (mn_avg - x_avg)
+
+        # assign the magnitudes for positive and negative anomalies
+        # convert to numpy masked arrays and then back out to xarray
+        rai = self.assign_magnitudes(y, anom, rai_plus, rai_minus, variable)
+
+        return rai
+
+    def fit(self, variable: str,
+            time_period: str = 'month',
+            rolling_window: int = 3) -> None:
+
+        print("Fitting Rainfall Anomaly Index")
+
+        # 1. calculate a cumsum over `rolling_window` timesteps
+        ds_window = rolling_cumsum(self.ds, rolling_window)
+
+        out_variable = 'RAI'
+        kwargs = {'variable': variable}
```

Comment:
this is one of the place causing the error:

```
src/analysis/indices/anomaly_index.py:124: error: Argument 6 to "apply_over_period" has incompatible type "**Dict[str, str]"; expected "Dict[Any, Any]"
```

which also relates to:
```
src/analysis/indices/utils.py:24: error: Incompatible types in assignment (expression has type "str", target has type "Dict[Any, Any]")
```

Reply:
Is this necessary? You create the `kwargs` dict only to pass it once to `apply_over_period`.

Is it possible you want to pass `kwargs`, not `**kwargs` in line 130?

Reply:
yes updated!